### DISCUSSION TITLE
Too slow search speed and invariable recall

### DISCUSSION THREAD

**Discussion by Molly-Xu00** [milvus-io/milvus#30584] [Q&A and General discussion]
I tried to conduct a hybrid search on a standalone milvus, the data vectors number is 1000000 (128 dim) and query number is 1000, I built HNSW index on data vectors and changed the search parameter 'ef' during search. I expected the recall-qps tradeoff would change as 'ef', however, the recall was always 100% and qps was always around 3. The qps was too slow and both recall and qps remained unchanged. Why did it come to such results?

**Comment by yhmo** [comment]
1. Set "Eventually" consistency level for search() interface, to reduce search latency.
collection.search(
  ...
  consistency_level="Eventually",
)


2. If you have filtering expression, the filtering expression might affect search performance for HNSW index. The fewer entities match the filtering condition, the poorer performance for HNSW index.

**Reply by Molly-Xu00** [reply]
Thank you so much! The search speed does improve. But the recall of search outcomes is always 100% and the QPS remains between 140~145. I hope that recall and QPS will change as I change the search parameter 'ef' ( when 'ef' becomes small, recall decreases and QPS increases), how can I get this?

**Reply by yhmo** [reply]
Theoretically, ef slightly affect recall rate and search performance. But if your data is small, I don't think you can feel the difference. Try large dataset that contains millions of entities.
If the collection is not fully indexed, brute-force search will be applied to segments that no index. Call collection.create_index() again to ensure all the segments are indexed.
ef range is [topk, 32768], your topk is 1000, which is a large value. You can set topk to 1, and set ef to 1 and try again.
